go/enclave/enclave.go

```diff
@@ -600,6 +609,160 @@ func (e *enclaveImpl) GetLogs(encryptedParams common.EncryptedParamsGetLogs) (co
+	// Binary search the gas requirement, as it may be higher than the amount used
+	var (
+		lo  = params.TxGas - 1
+		hi  uint64
+		cap uint64 //nolint:predeclared
+	)
+	// Use zero address if sender unspecified.
+	if args.From == nil {
+		args.From = new(gethcommon.Address)
+	}
+	// Determine the highest gas limit can be used during the estimation.
+	if args.Gas != nil && uint64(*args.Gas) >= params.TxGas {
+		hi = uint64(*args.Gas)
+	} else {
+		// TODO review this with the gas mechanics/tokenomics work
+		/*
+			//Retrieve the block to act as the gas ceiling
+			block, err := b.BlockByNumberOrHash(ctx, blockNrOrHash)
+			if err != nil {
+				return 0, err
+			}
+			if block == nil {
+				return 0, errors.New("block not found")
+			}
+			hi = block.GasLimit()
+		*/
+		hi = e.chain.GlobalGasCap
+	}
+	// Normalize the max fee per gas the call is willing to spend.
+	var feeCap *big.Int
+	if args.GasPrice != nil && (args.MaxFeePerGas != nil || args.MaxPriorityFeePerGas != nil) {
+		return 0, errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
+	} else if args.GasPrice != nil {
+		feeCap = args.GasPrice.ToInt()
+	} else if args.MaxFeePerGas != nil {
+		feeCap = args.MaxFeePerGas.ToInt()
+	} else {
+		feeCap = gethcommon.Big0
+	}
+	// Recap the highest gas limit with account's available balance.
+	if feeCap.BitLen() != 0 { //nolint:nestif
+		balance, err := e.chain.GetBalanceAtBlock(*args.From, blockNr)
+		if err != nil {
+			return 0, fmt.Errorf("unable to fetch account balance - %w", err)
+		}
+
+		available := new(big.Int).Set(balance.ToInt())
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(available) >= 0 {
+				return 0, errors.New("insufficient funds for transfer")
+			}
+			available.Sub(available, args.Value.ToInt())
+		}
+		allowance := new(big.Int).Div(available, feeCap)
+
+		// If the allowance is larger than maximum uint64, skip checking
+		if allowance.IsUint64() && hi > allowance.Uint64() {
+			transfer := args.Value
+			if transfer == nil {
+				transfer = new(hexutil.Big)
+			}
+			e.logger.Warn("Gas estimation capped by limited funds", "original", hi, "balance", balance,
+				"sent", transfer.ToInt(), "maxFeePerGas", feeCap, "fundable", allowance)
+			hi = allowance.Uint64()
+		}
+	}
+	// Recap the highest gas allowance with specified gascap.
+	if gasCap != 0 && hi > gasCap {
+		e.logger.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
+		hi = gasCap
+	}
+	cap = hi
+
+	// Create a helper to check if a gas allowance results in an executable transaction
+	executable := func(gas uint64) (bool, *gethcore.ExecutionResult, error) {
+		args.Gas = (*hexutil.Uint64)(&gas)
+		result, err := e.chain.ExecuteOffChainTransactionAtBlock(args, gethrpc.BlockNumber(0))
+		if err != nil {
+			if errors.Is(err, gethcore.ErrIntrinsicGas) {
+				return true, nil, nil // Special case, raise gas limit
+			}
+			return true, nil, err // Bail out
```

Comment:
Worth mentioning somewhere what the first return value in this helper function represents, I guess it's "failed"?

Maybe we could move this function out of its outer method to keep it tidy, make it `e.testGasAllowance` if possible or if it has to be constructed then maybe `testGasAllowance 
 = e.CreateGasAllowanceTester(...)`

Reply:
Good point, added a comment and moved the method out 👍